### PR TITLE
Add name searching functionality

### PR DESCRIPTION
**Pull request by Mizuko**
/listprofiles and /showprofile are great and all, but who uses GUIDs to actually look for things? People are bad databases and are a lot better at remembering reasonable strings... like names. This will add an optional name parameter to /listprofiles and /showprofile that filters down the profiles found by matching names like the given parameter. It also fixes /showprofile to show a message on no profiles found and only show the dropdown if more than one profile is found.

### CODE REVIEW

File: imageGenSD.py [hunk 1]
```diff
@@ -630,30 +661,54 @@ async def showprofile(interaction : dis.Interaction,
...
+                await interaction.response.send_message(content=f"<@{interaction.user.id}>", embed=embed)
+
+            elif len(profiles) == 1:
+
+                dis_log.debug(f"--Found one {name} {profiles[0].id}")
+                opts = {'id' : profiles[0].id}
+
+                dis_log.debug(f"Creating a job with metadata {metadata} and options {opts}.")
+                job = jf.JobFactory.getJob(type    = jf.JobTypeEnum.SHOW_PROFILE,
+                                        ctx     = interaction,
```
Comment (sBuiltaJr) [inline imageGenSD.py:695]: nit: vertical alignment is off for the last 2 parameters (maybe a tab snuck in somewhere)?

File: imageGenSD.py [hunk 2]
```diff
@@ -630,30 +661,54 @@ async def showprofile(interaction : dis.Interaction,
...
+                profiles = db_ifc.getProfiles(user_id, name)
+                empty_error = f'A character matching the name "{name}" wasn\'t found!'
+                many_message = f'Select a profile with name like "{name}" to view:'
 
-        await interaction.response.send_message(f'{result}', ephemeral=True, delete_after=9.0)
+            if len(profiles) == 0:
+
+                dis_log.debug(f"--Found none {name}")
+                embed = dis.Embed(title="Error",
+                                description=empty_error,
```
Comment (sBuiltaJr) [inline imageGenSD.py:684]: nit: vertical alignment is off for the last 2 parameters (maybe a tab snuck in somewhere)?

File: src/db/MariadbIfc.py [hunk 3]
```diff
@@ -400,14 +399,14 @@ def getProfile(self,
 
         return profile
 
-    def getProfiles(self,
+    def getUsersProfiles(self,
```
Comment (sBuiltaJr) [inline src/db/MariadbIfc.py:402]: nit: functions should be listed in alphabetical order, except for 'main' (or its equivalent) at the bottom.

File: imageGenSD.py [hunk 4]
```diff
@@ -630,30 +661,54 @@ async def showprofile(interaction : dis.Interaction,
...
-                                job      = job)
+                dis_log.debug(f"Looking for profiles of name {name}.")
+                profiles = db_ifc.getProfiles(user_id, name)
+                empty_error = f'A character matching the name "{name}" wasn\'t found!'
+                many_message = f'Select a profile with name like "{name}" to view:'
 
-        await interaction.response.send_message(f'{result}', ephemeral=True, delete_after=9.0)
+            if len(profiles) == 0:
+
+                dis_log.debug(f"--Found none {name}")
```
Comment (sBuiltaJr) [inline imageGenSD.py:682]: Debug message formats should be consistent with other comments (i.e. no '--' in the message).  this occurs in a few other places as well.

File: src/db/MariadbIfc.py [hunk 5]
```diff
@@ -400,14 +399,14 @@ def getProfile(self,
...
                     user_id : int) -> list:
```
Comment (sBuiltaJr) [inline src/db/MariadbIfc.py:403]: nit: vertical alignment is off for the last 2 parameters (maybe a tab snuck in somewhere)?

File: src/db/MariadbIfc.py [hunk 6]
```diff
@@ -427,6 +426,34 @@ def getProfiles(self,
...
         return results
 
+    def getProfiles(self,
+                    user_id : int,
+                    name : str) -> list:
+        """Returns profiles matching a given name filter for a given user.
+
+            Input: self - Pointer to the current object instance.
+                   user_id - user ID to interrogate for profiles.
+
```
Comment (sBuiltaJr) [inline src/db/MariadbIfc.py:436]: Missing parameter description for 'name'.

File: src/db/MariadbIfc.py [hunk 7]
```diff
@@ -427,6 +426,34 @@ def getProfiles(self,
...
+            Output: list - A list of all profiles found, if any.  An empty list if not.
+        """
+        cmd     = ""
+        cursor  = self.con.cursor(buffered=False)
+        results = []
+
+        self.db_log.info(f"Getting profiles for user {user_id}")
```
Comment (sBuiltaJr) [inline src/db/MariadbIfc.py:443]: The log message should be unique so its easy to tell which function was called from the log alone.  Maybe something like "Searching for profiles matching {name} for {user_id}"?

File: imageGenSD.py [hunk 8]
```diff
@@ -597,16 +609,19 @@ async def roll(interaction: dis.Interaction):
 
 @IGSD_client.tree.command()
 @dac.checks.has_permissions(use_application_commands=True)
-@dac.describe(profile_id="The profile ID of the character you'd like to view.  Use /listprofiles to see the name and ID for profiles!")
-@dac.describe(user="The Discord user owning the profiles listed by the command.  If none, it defaults to you.")
+@dac.describe(profile_id="The profile ID of the character.  Use /listprofiles to find the ID.")
+@dac.describe(user="The Discord user owning the profiles.  If none, it defaults to you.")
+@dac.describe(name="A name of the profile to view. Will find all similar names.")
```
Comment (sBuiltaJr) [inline imageGenSD.py:614]: This should mention that the search is case-insensitive.

File: imageGenSD.py [hunk 9]
```diff
@@ -369,13 +373,21 @@ async def listprofiles(interaction : dis.Interaction,
     #This has to be in the function body because an arg can't be used to assign
     #another arg in the function call.
     user_id = interaction.user.id if user == None else user.id
-    profiles = db_ifc.getProfiles(user_id)
+    user_dis = interaction.user if user == None else user
+    error_desc = ""
+    profiles = []
+    if name == None:
+        profiles = db_ifc.getAllProfiles(user_id)
```
Comment (sBuiltaJr) [inline imageGenSD.py:380]: This appears to be the old function name and should be getUserProfiles (?)

**Review by sBuiltaJr — COMMENTED** [review]
Mostly nits, but one definite actionable change.  Review to be completed at a later time.

File: imageGenSD.py (hunk 2 above)
Comment (Mizuko) [reply imageGenSD.py]: nah, it looks like not paying attention to the auto-indent. I think you missed the same misalignment in the next part in the elif

**Review by sBuiltaJr — APPROVED** [review]

**Pull request merged**